main/main.go

```diff
@@ -0,0 +1,298 @@
+				return cli.NewExitError("Error: cannot add or remove keys on non-existent files, use `--kms` and `--pgp` instead.", 49)
+			}
+			if c.Bool("encrypt") || c.Bool("decrypt") || c.Bool("rotate") {
+				return cli.NewExitError("Error: cannot operate on non-existent file", 100)
+			}
+		}
+		fileBytes, err := ioutil.ReadFile(file)
+
+		var output *os.File
+		if c.Bool("in-place") {
+			var err error
+			output, err = os.Create(file)
+			if err != nil {
+				return cli.NewExitError(fmt.Sprintf("Could not open in-place file for writing: %s", err), 10)
+			}
+			defer output.Close()
+		} else {
+			output = os.Stdout
+		}
+		if err != nil {
+			return cli.NewExitError("Error: could not read file", 2)
+		}
+		if c.Bool("encrypt") {
+			return encrypt(c, file, fileBytes, output)
+		} else if c.Bool("decrypt") {
+			return decrypt(c, file, fileBytes, output)
+		} else if c.Bool("rotate") {
+			return rotate(c, file, fileBytes, output)
+		} else {
+
+		}
+		return nil
+	}
+	app.Run(os.Args)
+}
+
+func runEditor(path string) {
+	editor := os.Getenv("EDITOR")
+	var cmd *exec.Cmd
+	if editor == "" {
+		cmd := exec.Command("which", "vim", "nano")
+		out, err := cmd.Output()
+		if err != nil {
+			panic("Could not find any editors")
+		}
+		cmd = exec.Command(strings.Split(string(out), "\n")[0], path)
+	} else {
+		cmd = exec.Command(editor, path)
+	}
+	cmd.Run()
+}
+
+func store(path string) sops.Store {
+	if strings.HasSuffix(path, ".yaml") {
+		return &yaml.YAMLStore{}
+	} else if strings.HasSuffix(path, ".json") {
+		// return &json.JSONStore{}
+	}
+	panic("Unknown file type for file " + path)
+}
+
+func findKey(keysources []sops.KeySource) (string, error) {
+	for _, ks := range keysources {
+		for _, k := range ks.Keys {
+			key, err := k.Decrypt()
+			if err == nil {
+				return key, nil
+			}
+		}
+	}
+	return "", fmt.Errorf("Could not get master key")
+}
+
+func decrypt(c *cli.Context, file string, fileBytes []byte, output io.Writer) error {
+	store := store(file)
+	metadata, err := store.LoadMetadata(string(fileBytes))
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Error loading file: %s", err), 3)
+	}
+	key, err := findKey(metadata.KeySources)
+	if err != nil {
+		return cli.NewExitError(err.Error(), 4)
+	}
+	branch, err := store.Load(string(fileBytes))
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Error loading file: %s", err), 6)
+	}
+	tree := sops.Tree{Branch: branch, Metadata: metadata}
+	mac, err := tree.Decrypt(key)
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Error decrypting tree: %s", err), 8)
+	}
+	if metadata.MessageAuthenticationCode != mac && !c.Bool("ignore-mac") {
+		return cli.NewExitError("MAC mismatch.", 9)
+	}
+	out, err := store.Dump(tree.Branch)
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Error dumping file: %s", err), 7)
+	}
+	_, err = output.Write([]byte(out))
```

Comment:
did `out` ever need to not be a `[]byte` ? If not, just return it as an `[]byte` right away.